Improve under/overflow and nan handling in binning related functions

maxnoe opened the pull request:
Underflow and overflow are now marked with special values (intmin, intmax) that should prevent silent mistakes such as indexing the last entry because the underflow index was -1 before. Using undeflown values to index into a histogram now correctly raises an index error.

The benchmarks / calculate_percentile_cut functions no use the `nan<>` family of functions, so they ignore nan values in their input.

TarekHC commented:
I can confirm this PR was working properly once I changed the following:
```python
table = QTable({"values": values, "bin_values": bin_values}, copy=False)
```
to:
```python
table = QTable({"values": values, "bin_values": bin_values}, copy=True)
```
I tested it comparing eventDisplay "cut class 2" events sensitivity vs pyirf sensitivity, showing perfect matching.

maxnoe commented:
@TarekHC I opened https://github.com/astropy/astropy/issues/13840 and will add a workaround here

JBernete reviewed (dismissed):
Every thing looks good to me

maxnoe commented:
@JBernete I missed a bug (you too ;) ) where I confused quantile and percentile of an untested function.

I fixed it and added a test case. Please review again.

JBernete commented:
@maxnoe I see, I agree with the change. I'm not still experienced with checks and I see there is a warning that some checks were not succesful. How important is this for the PR?

maxnoe commented:
@JBernete The docs build fail is crucuial, that needs to pass. The coverage is "nice to have" and should not be too bad.

maxnoe commented:
The docs build seems to have been a transient connection problem downloading something
